Documentation/botr/readytorun-overview.md

```diff
@@ -0,0 +1,335 @@
+
+### Performance Impact
+
+Clearly we have added an instruction and thus made the code bigger and more expensive to run. However what is also true is that the additional cost is small. The 'worst' case would be if this field fetch was in a tight loop. To measure this we created a linked list element which inherited across a version bubble. The list was long (1K) but small enough to fit in the L1 cache. Even for this extreme example (which by the way is contrived, linked list nodes do not normally inherit in such a way), the extra cost was small (< 1%).
+
+### Null checks 
+
+The managed runtime requires any field access on null instance pointer to generate null reference exception. To avoid inserting explicit null checks, the code generator assumes that memory access at addresses smaller than certain threshold (64k on Windows NT) will generate null reference exception. If we allowed unlimited growth of the base class for cross-version bubble inheritance hierarchies, this optimization would be no longer possible.
+
+To make this optimization possible, we will limit growth of the base class size for cross-module inheritance hierarchies. It is a new versioning restriction that does not exist in IL today.
+
+
+## Non-Virtual Method Calls - CALL
+
+### Intra-module call
+ 
+If RCX holds a `BaseClass` and the caller of `iMethod` is in the same module as BaseClass then a method call is simple machine call instruction
+
+        CALL ENTRY_IMETHOD
+
+### Inter-module call
+ 
+However if the caller is outside the module of BaseClass (even if it is in the same versioning bubble) we need to call it using an indirection
+
+	    CALL [PTR_IMETHOD]
+
+	.data // In the data section
+    PTR_IMETHOD: PTR = RUNTIME_ENTRY_FIXUP_METHOD // One per call TARGET. 
+
+Just like the field case, the pointer sized data slot `PTR_IMETHOD` must be fixed up to point at the entry point of `BaseClass.iMethod`. However unlike the field case, because we are fixing up a call (and not a MOV), we can have the call fix itself up lazily via standard delay loading mechanism.
+The delay loading mechanism often use low-level tricks for maximum efficiency. Any low-level implementation of delay loading can be used as long the resolution of the call target is left to the runtime.
```

Comment:
~~often use low-level~~ often *uses* low-level
~~can be used as long the~~ can be used as long *as* the